sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java

```diff
@@ -0,0 +1,474 @@
+    }
+
+    public Write withMetastoreUri(String metastoreUri) {
+      return toBuilder().setMetastoreUri(metastoreUri).build();
+    }
+
+    public Write withDatabase(String database) {
+      return toBuilder().setDatabase(database).build();
+    }
+
+    public Write withTable(String table) {
+      return toBuilder().setTable(table).build();
+    }
+
+    public Write withFilter(Map filter) {
+      return toBuilder().setFilter(filter).build();
+    }
+
+    public Write withBatchSize(long batchSize) {
+      return toBuilder().setBatchSize(batchSize).build();
+    }
+
+    @Override
+    public PDone expand(PCollection<HCatRecord> input) {
+      input.apply(ParDo.of(new WriteFn(this)));
+      return PDone.in(input.getPipeline());
+    }
+
+    @Override
+    public void validate(PipelineOptions options) {
+      checkNotNull(getMetastoreUri(), "metastoreUri");
+      checkNotNull(getTable(), "database");
+    }
+
+    private static class WriteFn extends DoFn<HCatRecord, Void> {
+      private final Write spec;
+      private WriterContext writerContext;
+      private HCatWriter slaveWriter;
+      private HCatWriter masterWriter;
+      private List<HCatRecord> hCatRecordsBatch;
+
+      public WriteFn(Write spec) {
+        this.spec = spec;
+      }
+
+      @Override
+      public void populateDisplayData(
+          org.apache.beam.sdk.transforms.display.DisplayData.Builder builder) {
+        super.populateDisplayData(builder);
+        builder.add(DisplayData.item("database", spec.getDatabase()));
+        builder.add(DisplayData.item("table", spec.getTable()));
+        builder.add(DisplayData.item("filter", (spec.getFilter() != null
+            ? spec.getFilter().toString() : "NULL")));
+        builder.add(DisplayData.item("metastoreUri", spec.getMetastoreUri()));
+        builder.add(DisplayData.item("batchSize", spec.getBatchSize()));
+      }
+
+      @Setup
+      public void initiateWrite() throws HCatException {
+        try {
+          //initiate writerContext, masterWriter & slaveWriter
+          //to be used for writing into Hive
+          writerContext = getWriterContext();
+          slaveWriter = DataTransferFactory.getHCatWriter(writerContext);
+        } catch (HCatException e) {
+          LOG.error("Exception in initiateWriterContext", e);
+          throw e;
+        }
+      }
+
+      @StartBundle
+      public void startBundle() {
+        hCatRecordsBatch = new ArrayList<HCatRecord>();
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext ctx) throws HCatException {
+        hCatRecordsBatch.add(ctx.element());
+        //apply batch
+        if (hCatRecordsBatch.size() >= spec.getBatchSize()) {
+          flush();
+        }
+      }
+
+      @FinishBundle
+      public void finishBundle() throws HCatException {
+        flush();
+      }
+
+      private void flush() throws HCatException {
+        try {
+          //write data using slaveWriter
```

Comment:
These comments seem unnecessary because they don't communicate any information that is not already apparent from the code.